### ISSUE TITLE
PossiblyInvalidArgument false-positive for Collection methods

### ISSUE THREAD

**Issue by fluffycondor** [weirdan/doctrine-psalm-plugin#92]
vimeo/psalm 4.8.1
weirdan/doctrine-psalm-plugin 1.0.0

```php
/** @param Collection<int, A>|Collection<int, B> */
public function deleteItemById(Collection $collection, int $id): void
{
    $matchedItems = $collection->filter(
        static function ($inCollectionItem) use ($id): bool {
            return $inCollectionItem->haveId($id);
        }
    );

    if (!$matchedItems->isEmpty()) {
        $collection->removeElement($matchedItems->first()); // yields PossiblyInvalidArgument
    }
}
```

`psalm: PossiblyInvalidArgument: Argument 1 of Doctrine\Common\Collections\Collection::removeElement expects A, possibly different type A|B provided`

It looks false-positive to me, it's impossible to call here `Collection<int, A>::removeElement(B)`.
